Searchable Attributes behaviors (#3)
meilisearch/specifications: text/0002-searchable-display-attributes-behaviors.md

- > **is not in `searchable_attributes` but in `displayed_attributes`:**
  > meilisearch will **not** use the field for a search but the field will be returned in search results
- > **is not in `searchable_attributes` and not in `displayed_attributes`:**
  > meilisearch will **not** use the field for a search and the field will **not** be returned in search results. The field is unused but should be kept in case of it is added in `searchable_attributes` or `displayed_attributes`.

#### add a new configuration `attributes_order`?

This new config may define the order of attributes for the criterion `Attributes`, no wildcard allowed.
if `searchable_attribute` is: `["title", "desc"]`
and `attributes_order` is: `["desc", "id", "title"]`
The searchable attributes order would be:
`"desc" > "title"` (`"id"` is not searchable so is ignored by the criterion `Attributes`).

`attributes_order` is always exhaustive, If a new field is added, it is "pushed" at the end of the list.

##### drawbacks:
- new config added
- more complexity
- could be replace by a small note about wildcard in `searchable_attributes`

#### Changes in meilisearch documentation

> Depending on what we choose, we'll have to add **default** and **wildcard** behavior of configurations, we'll potentially add a new configuration `attributes_order`. 

Reviewer: I would like us to consider the possibility of removing the wildcard, as it seems "simple" for usage at first sight, but introduces complexity as behavior is not clear and needs to be documented/learned, and is subject to confusion.

This also would make it possible to use just one setting, stating CLEARLY that the order MATTERS! So it could solve both issues without adding a new setting!

> Depending on what we choose, we'll have to add **default** and **wildcard** behavior of configurations, we'll potentially add a new configuration `attributes_order`. 

Reviewer 2: This is not an option. We decided to add the wildcard because without it we needed acceptNewField whose use needed an understanding of how meilisearch used new fields and was not clear at all

> Depending on what we choose, we'll have to add **default** and **wildcard** behavior of configurations, we'll potentially add a new configuration `attributes_order`. 

Reviewer: The more I reflect on this, the more I feel the need to create `attributes_order` as in this proposition. I would try hard to avoid adding a new setting, but it seems to be the only way (for now) to offer logic, consistent and clear behavior for this issue.

> Depending on what we choose, we'll have to add **default** and **wildcard** behavior of configurations, we'll potentially add a new configuration `attributes_order`. 

Reviewer: > I think that MeiliSearch being easy to use is more about the fact that you have very little configuration to do to make it work in most case than the fact that we have "few settings".

I think it is a mix of both

> Depending on what we choose, we'll have to add **default** and **wildcard** behavior of configurations, we'll potentially add a new configuration `attributes_order`. 

Reviewer 3: I thought a lot about this, and I also feel really concerned about the implicit-and-not-obvious role of `searchableAttributes`, i.e the importance order of the fields.

And in this case, adding another settings might be the best option to manage the importance of the fields during the search.

Let's call this new setting `XXX`.

Here what we should be aware about if we add this new settings:
- Adding a new setting is adding an obstacle for the user if there are too many settings, even if each settings are explicit and well documentated. I had bad experience the first time I tried Algolia and I saw all the possibilities. No need to detail my experience with Elastic even if all the Elastic settings are really accurate and the docs not that bad. So, yes, adding a settings is adding complexity from my POV. We should be carefull about settings addition, but it does not mean not adding another settings at all. If `XXX` is needed, we should add it.
- The name of `XXX` should be really accurate
- We should anticipate, at the maximum, all the behaviors between `searchableAttributes` and `XXX`. What's about if there are less fields in `searchableAttributes` than in `XXX`?
- We should anticipate all the behavior possibilities when the users try to update `XXX`: if the list of attributes is not exhaustive, does MeiliSearch return an error or do an implicit behavior?
- How the `XXX` is going to be filled if the users add documents with new fields?

If we don't add `XXX` the docs should be really well explained and the users should be able to understand as soon as possible the `searchableAttributes` settings allow them to do two things.

The more I'm thinking about this, the more I would like to add a new setting, but I don't have any strong position about this. And if we add this new settings, we should be careful about a lot of behaviors.

> Depending on what we choose, we'll have to add **default** and **wildcard** behavior of configurations, we'll potentially add a new configuration `attributes_order`. 

Reviewer 3: And we definitely need to talk directly about this to take a decision: adding a new settings or not 🙂

> Depending on what we choose, we'll have to add **default** and **wildcard** behavior of configurations, we'll potentially add a new configuration `attributes_order`. 

Reviewer: Seems like time has come for a meeting! :)

- Title: Searchable Attributes and Displayed Attributes behaviors
- Start Date: 2020-11-02
- specification PR: meilisearch/specifications#3
- Meilisearch Issue: 

# `searchable_attributes` & `displayed_attributes` behaviors

## First section: Feature Description and Interaction

This first part has a general audience. It should be as little technical as possible (think user-level). This section contains 4 sub-sections:

### Summary

> Ever since the beginning, the design of `searchableAttributes` and `displayedAttributes` as shown flaws. The introduction of the wildcard tried to adress some of those flaws, but there are still issues that this specs tries to correct.

Reviewer: ```suggestion
Ever since the beginning of MeiliSearch, the design of `searchableAttributes` and `displayedAttributes` has shown flaws. The introduction of the wildcard tried to address some of those flaws, but there are still issues that this spec tries to correct.
```

Searchable attributes have several uses:

> - know in which fields we can search.
> - know the importance of fields for the criterion `Attributes`.

Reviewer: ```suggestion
- Stablish in which document fields the search can be performed.
- Define the order and priority in which those fields are used by the criterion `Attributes`.
```

Reviewer 4: I think you meant `Establish` @eskombro

Reviewer: Danke ;)

> The main issues we want to adress here are:

Reviewer: ```suggestion
The main issues we want to address here are:
```

> - The question of `searchableAttributes` order when the wildcard is used.

Reviewer: ```suggestion
- The behavior of `searchableAttributes` order when the wildcard is used.
```

> - The fact that already known attributes (e.g from facets), are not indexed correctly is put before the documents (bug), need to be taken into account in the design.

Reviewer: ```suggestion
- The fact that already known attributes (e.g from facets), are not indexed correctly if the `searchableAttributes` setting is defined before the documents indexation (bug), needs to be taken into account in the design.
```

Is it relevant to mix this 2 information when we have wildcards?

### Explanation

#### default values

`searchable_attributes`: *   

> `displayed_attributes`: *

Reviewer: We should choose a consistent syntax: `searchableAttributes` or `searchable_attributes` :)

#### wildcard (*) & `searchable_attributes` order

> > When the searchable attributes value is set to ["*"] the priority of the documents attributes is undefined.

Reviewer: We should choose a consistent syntax: `*` or `["*"]` :)

By default searchable attributes should be in the order of the first indexed document, for example:
```json
{
  "title": "many the fish",
  "description": "it is awfull to be liked by many"
}
```
default `searchable_attributes` order should be `["title", "description"]` where `"title" > "description"`

When a wildcard is set we should have the default behavior.

#### `searchable_attributes` & `displayed_attributes`

what is the behavior for a field that:
- > **is in `searchable_attributes` and in `displayed_attributes`:**
  > meilisearch will use the field for a search and the field will be returned in search results
- > **is in `searchable_attributes` but not in `displayed_attributes`:**
  > meilisearch will use the field for a search but the field will **not** be returned in search results
- > **is not in `searchable_attributes` but in `displayed_attributes`:**
  > meilisearch will **not** use the field for a search but the field will be returned in search results
- > **is not in `searchable_attributes` and not in `displayed_attributes`:**

>   > meilisearch will **not** use the field for a search and the field will **not** be returned in search results. The field is unused but should be kept in case of it is added in `searchable_attributes` or `displayed_attributes`.

Reviewer: ```suggestion
  > meilisearch will **not** use the field for a search and the field will **not** be returned in search results. The field is unused but should be kept in case of is added in `searchable_attributes` or `displayed_attributes`, or present in a new document.
```

#### add a new configuration `attributes_order`?

> This new config may define the order of attributes for the criterion `Attributes`, no wildcard allowed.

Reviewer: ```suggestion
This new Setting may define the order of attributes for the criterion `Attributes`, no wildcard allowed.
```

Reviewer 4: maybe the capital letter on the setting is not needed? Apart from that, I agree.

if `searchable_attribute` is: `["title", "desc"]`
and `attributes_order` is: `["desc", "id", "title"]`
The searchable attributes order would be:
`"desc" > "title"` (`"id"` is not searchable so is ignored by the criterion `Attributes`).

`attributes_order` is always exhaustive, If a new field is added, it is "pushed" at the end of the list.

##### drawbacks:
- new config added
- more complexity

> - could be replace by a small note about wildcard in `searchable_attributes`

Reviewer: ```suggestion
- could be replaced by a small note about wildcard in `searchable_attributes`
```

Reviewer: This idea needs some development, it isn't clear at all what it means to replace this with a 'small note' :)

Reviewer 2: domumenting strange behaviour != fixing strange behaviour, this is a non-solution to me. @bidoubiwa will tell you that people don't really read the doc in depth the first time around. What we want is discoverability + intuitiveness

Reviewer 4: I think you tend to read the doc about a specific subject if you think you don't understand it. In the case of `searchableAttribute` the meaning is very clear in its first use `in what attributes will meilisearch search`. Because that meaning is so intuitive, you wouldn't guess that there is a second one to look after in the documentation.

So even if they in fact read the documentation, they would probably just read over that part because they think they already understand it. 

This is speculation of course, maybe this is not how users read. This is how I read and how I think people read.

Reviewer: Oh, men, I am in your team now, @MarinPostma and @bidoubiwa !

- If applicable, provide sample error messages, deprecation warnings, or migration guidance.

If the changes modify the HTTP API, provide a description of the method, URL, parameters, body, status code, errors, etc...

If it modifies the CLI, provide the env variable name, the argument name, and the description.

This serves as a user-level guide. Anything that the user may encounter during its interaction with the feature should be presented here.
Impact on documentation

If the feature requires additions to the documentation or if sections of the documentation need to be updated because of this feature, it should be mentioned here. It's the role of the documentation team to point out the sections of the documentation that need to be updated. -->

## Second Section: Technical Specifications

<!-- This section has a much narrower audience: the developer that will implement the feature. Its goal is to make it as clear as possible to develop the feature, share knowledge, and think about the possibilities. -->

### Architecture

> This change will impact **settings**, the **indexer** and the **search-engine** (`Attribute` criterion).

Reviewer: ```suggestion
This change will impact **settings**, the **indexer** and the **search-engine** (`Attributes` criterion).
```

Reviewer 4: Should be lowercase if put between backticks

This means that there should be a distinction between the creation of an `fieldId` (when a field is first registered), and it's usage/position in `searchableAttributes` and `displayedAttributes`.
What that means is that when we index documents, we must check if a field already has a position, and if not, give it a position (after adding it to the fieldmap) according to its position in the document.

##### caveat
- If a new document contains a new field at a position for which we already have a field:
```json
[
	{
		"id": 1,
		"title": "the title"
	},
	{

		"id": 1,
		"titre": "le titre"
	}
]
```

- To make sure that the `searchableAttributes` and `displayedAttributes` are in sync, the must come from the same source of information; this should be the field_position.
 TBD: What should be the position of `titre`?
	- same position as `title` (thus allowing attributes to exists at the same position like Algolia)
	- added at the last position

#### `searchableAttributes` & `displayedAttributes`

Being searchable and displayed are 2 different concept, and there should not be interactions between the two: a field can be searchable and not displayed, or displayed and not searchable, etc.

> #### add a new configuration `attributesPosition`?

Reviewer 4: Are we sure about this name? Some other suggestions: 

- `attributesImportanceOrder`
- `attributesRelevancyOrder`
- `attributesOrderOfRelevancy`
- ...

I like a clear name. I feel like `attributesPosition` is not clear in it's utility.

Reviewer 2: After some research on the technical feasibility of the `attributePostition` (whatever we decide to call it in the future), I have found that it is a little bit more complicated than anticipated. 

We don't currently support attributes having the same importance ([algolia does it](https://www.algolia.com/doc/guides/managing-results/must-do/searchable-attributes/how-to/configuring-searchable-attributes-the-right-way/#searchable-attributes-with-the-same-priority)). This is a pre-requisite for this to work, and this is not easy to do with the current version (nothing impossible in my mind though).

Now, we will probably want to support attributes with the same priority at some point in time (I see no reason why we wouldn't) and this should be taken into account in the current design AND future design of MS (@Kerollmops).

### Technical reason

In order to know what position a field is in, the `DocIndex` struct contains an `attributes` which instead of being a `FieldId` is the position. This system is designed with the the position being isomorphic to the attibute id, so one position maps to only one attribute. If we allow multiple attributes to have the same importance, then we can't rely on that.
we could either:
- store the position int he `DocIndex`
- store the fieldId in the `DocIndex` and map it to a position at a later time

Adding a new configuration `attributesPosition` that link all attributes present in a document  to a position makes it possible to keep track of the position without caring about whether a field is searchable or displayed. This also allows to keep track of all the fields that have been found in documents in the database.

The default order is then **always** the order a field was in a document when it was first found.

> ##### Content of `attributesPosition`

Reviewer 4: The absence of a field in `searchableAttributes` does not impact what will be added to `attributesPosition`?

Reviewer 2: no it doesn't, all fields that are present in any documents are given an attribute position. This way you can add or remove an attribute position, and this will not impact it's position in the documents. No data is ever lost.

e.g: In this example is makes sense that this is the default:

```json
/// documents
[
	{
		"id": 1,
		"title": "the title"
	},
	{

		"id": 2,
		"titre": "le titre"
	}
]
```
```json
/// attributesPosition
{
	"id": 0,
	"title": 1,
	"titre": 1
}
```

##### Updating the `attributesPosition`

There are 2 kinds of operations we need to consider:

1) **Deletion**: `attributesPosition` are essential to the functionning of meilisearch. Their deletion should not be allowed. However they can be cleaned:
calling `DELETE /indexes/:index_uid/attributes_position` triggers a reindexation, and only the fields present in the documents are kept. e.g:
if my `attributesPosition` look like this: 
```json
/// attributesPosition
{
	"id": 0,
	"title": 1,
	"titre": 1
}
```
and thoses are all of my documents are:
```json
/// documents
[
	{
		"id": 1,
		"title": "the title"
	},
	{

		"id": 2,
		"title": "the title2"
	}
]
```

then a call to `DELETE /indexes/:index_uid/attributes_position` will result in a re-indexation and `attributesPosition` will look like this: 
```json
/// attributesPosition
{
	"id": 0,
	"title": 1
}
```

> 2) **Updates**: As we have seen before, we don't want to allow deletion of those fields unless we are sure that this field does not exist in any documents. That would imply that trying to update the `attributesPosition` with a missing value is not allowed.

Reviewer 4: what is the maximum number? Can it have negative values? Can there be gaps between positions?

Reviewer 2: this is TBD, it will probably be bound by the type of the value we use to store the information, this is an implementation detail I don't have the answer to now. Gaps are tolerated.
